processors/locations.py

```diff
@@ -0,0 +1,169 @@
+       (at your option) any later version.
+       This program is distributed in the hope that it will be useful,
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    This module implements a processor that looks at parsed sentence trees
+    and extracts any addresses / locations, looks up information about
+    them and saves to a database.
+
+"""
+
+from collections import namedtuple
+from datetime import datetime
+from scraperdb import Location
+from geo import (
+    coords_for_country,
+    coords_for_street_name,
+    icelandic_addr_info,
+    isocode_for_country_name,
+    ICELAND_ISOCODE,
+)
+
+PLACENAME_BLACKLIST = ["Staður", "Eyjan", "Fjöll", "Bæir"]
+
+Loc = namedtuple("Loc", "name kind")
+
+
+def article_begin(state):
+    """ Called at the beginning of article processing """
+
+    # Delete all existing locations for this article
+    session = state["session"]  # Database session
+    url = state["url"]  # URL of the article being processed
+    session.execute(Location.table().delete().where(Location.article_url == url))
+
+    # Set of all unique locations found in article
+    state["locations"] = set()
+
+
+def article_end(state):
+    """ Called at the end of article processing """
+
+    url = state["url"]
+    session = state["session"]
+
+    locs = state.get("locations")
+    if not locs:
+        return
+
+    print(locs)
+
+    # Find all placenames mentioned in article
+    # We can use them to disambiguate addresses and street names
+    placenames = [p.name for p in locs if p.kind == "placename"]
+
+    # Get info about each location and save to database
+    for loctuple in locs:
+        loc = loctuple._asdict()
+        kind = loc["kind"]
+        coords = None
+
+        # Heimilisfang
+        if kind == "address":
+            loc["country"] = ICELAND_ISOCODE
+            info = icelandic_addr_info(loc["name"], placename_hints=placenames)
+            if info and info.get("lat_wgs84") and info.get("long_wgs84"):
+                coords = (info["lat_wgs84"], info["long_wgs84"])
+            loc["data"] = info
+
+        # Land
+        elif kind == "country":
+            code = isocode_for_country_name(loc["name"])
+            if code:
+                loc["country"] = code
+                coords = coords_for_country(code)
+
+        # Götuheiti
+        elif kind == "street":
+            # All the street names in BÍN are Icelandic
+            loc["country"] = ICELAND_ISOCODE
+            coords = coords_for_street_name(loc["name"], placename_hints=placenames)
+
+        # Örnefni
+        elif kind == "placename":
+            if loc["name"] in PLACENAME_BLACKLIST:
```

Comment:
Is upper and lower case definitely not significant here?

Reply:
Well, at this point we're working with the canonical nominative form of the placename (in fl "örn) from BÍN, which should be capitalised.

Reply:
But point taken, no harm in having the comparison case insensitive.